### PR TITLE
Feature/attribution toolkit

### PR DESCRIPTION
**Pull request by Sammac-dev**
# Description

- Attribution Toolkit, allows users to calculate and visualize both portfolio attribution by sector and benchmark attribution by sector for the S&P500.

To use attribution toolkit run "attrib" from portfolio section after loading in portfolio

image
image
image

Requested by Product Manager Jeroen Bouma

- Added dependency - yahooquery==2.2.10

# How has this been tested?
This feature has been tested locally by running the terminal and checking that the outputs display correctly and as expected. The feature is found within the portfolio menu and is called via "attrib". The feature has been executed with default time period of all (using first date found in the portfolio loaded by the user) via the "attrib" command and also with different time periods selected such as "attrib -p 3m". 

# Checklist:

- [X ] Update [our Hugo documentation](https://openbb-finance.github.io/OpenBBTerminal/) following [these guidelines](https://github.com/OpenBB-finance/OpenBBTerminal/tree/main/website).
- [x] Update our tests following [these guidelines](https://github.com/OpenBB-finance/OpenBBTerminal/tree/main/tests). - manual testing has been performed, unit testing has not 
- [X ] Make sure you are following our [CONTRIBUTING guidelines](https://github.com/OpenBB-finance/OpenBBTerminal/blob/main/CONTRIBUTING.md).
- [X ] If a feature was added make sure to add it to the corresponding [scripts file](https://github.com/OpenBB-finance/OpenBBTerminal/tree/main/scripts).

# Others
- [X ] I have performed a self-review of my own code.
- [X ] I have commented my code, particularly in hard-to-understand areas.

### CODE REVIEW

**Comment by JerBouma** [comment]
Awesome @Sammac-dev, I'll be deep diving into this also checking if the financial theory behind it is correct. Will get back to you soon 😁

**Comment by jmaslek** [comment]
Thanks @Sammac-dev.  This looks awesome!

One question: can you replace the yahoo query with yfinance.  Since you can get the sector weightings with `yf.Ticker("SPY").info["sectorWeightings"]`.  With a massive list of dependencies, I am inclined to stick to the package we have that already does this, just to not keep adding to that list.

Granted - wow this is faster
Screen Shot 2022-10-29 at 11 25 32 AM

**Comment by JerBouma** [comment]
Awesome stuff @Sammac-dev, I'll be looking into it combined with checking if the financial theory behind it checks out.

**Comment by Sammac-dev** [comment]
Thanks @JerBouma 
@jmaslek no problems I can remove the yahooquery dependencies and replace with yahoofinance. Also, as this is our first time contributing to an open-source project there are a couple of areas I'm unsure of. 
1. In order to send the updates to you do I simply commit to my feature/AttributionToolkit branch or do I need to do something else?

**Comment by jmaslek** [comment]
Yes you can just commit to your branch!

**Comment by Sammac-dev** [comment]
@jmaslek that's been taken care of, I've committed the new code

**Review by JerBouma — CHANGES_REQUESTED** [review]
Hi @Sammac-dev, I made quite some suggestions. The most prevalent ones are making sure that there are no calculations/operations done within the controller other than calling the model/view and that you segregate the attribution functionalities well enough so users can pick what they want. Feel free to contact me if you have any questions.

File: openbb_terminal/portfolio/sp500_sectors_and_industries.json [hunk 1]
```diff
@@ -0,0 +1,247 @@
+{
```
[the remark this answers is not in the document]
Comment (Sammac-dev) [reply openbb_terminal/portfolio/sp500_sectors_and_industries.json:1]: Hi Jeroen, we can hardcode a sector/ticker map into the model if seeing as we just have sector attrib at the moment.

[the remark this answers is not in the document]
Comment (Sammac-dev) [reply openbb_terminal/portfolio/sp500_sectors_and_industries.json:1]: for reference the decision to use JSON or hardcoding sector maps is due to the get_data.py function taking a long time to run and missing the ticker for energy sector.

File: requirements.txt [hunk 2]
```diff
@@ -1,328 +1,328 @@
-aiodns==3.0.0; python_version >= "3.5.2" and python_full_version >= "3.7.0"
```
[the remark this answers is not in the document]
Comment (Sammac-dev) [reply requirements.txt:1]: Yeah, I think that might have something to do with the changes on main whilst we're branched. I'll add requirements.txt to the gitignore as we have no changes for this file for the project now that we've removed yahoo query.

**Comment by Sammac-dev** [comment]
@JerBouma all the changes above have been committed. The only thing I'm not sure on is how to resolve the requirements.txt file. From our end we have no need for any changes to it outside of what is already on OpenBB-finance main branch. So let us know if there is anything we need to do from our end.

File: openbb_terminal/portfolio/sp500_sectors_and_industries.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (JerBouma) [reply openbb_terminal/portfolio/sp500_sectors_and_industries.json]: Sounds good to keep the JSON then!

File: openbb_terminal/portfolio/portfolio_controller.py [hunk 3]
```diff
@@ -638,6 +643,112 @@ def call_alloc(self, other_args: List[str]):
...
+                """,
+        )
+        parser.add_argument(
+            "-p",
+            "--period",
+            type=str,
+            choices=portfolio_helper.PERIODS,
+            dest="period",
+            default="all",
+            help="The file to be loaded",
```
Comment (JerBouma) [inline openbb_terminal/portfolio/portfolio_controller.py:664]: Note this should be about the period.

File: openbb_terminal/portfolio/portfolio_controller.py [hunk 4]
```diff
@@ -638,6 +643,112 @@ def call_alloc(self, other_args: List[str]):
...
+        )
+        parser.add_argument(
+            "-r",
+            "--raw",
+            type=bool,
+            dest="raw",
+            default=False,
+            const=True,
+            nargs="?",
+            help="view attribution in a graph visualisation format",
```
Comment (JerBouma) [inline openbb_terminal/portfolio/portfolio_controller.py:683]: With `-r` you show the table + graph, this does not disable the graph.

**Review by JerBouma — CHANGES_REQUESTED** [review]
Some minor details and then it should be finished! :)

File: openbb_terminal/portfolio/portfolio_controller.py (hunk 3 above)
Comment (Sammac-dev) [reply openbb_terminal/portfolio/portfolio_controller.py:664]: Hi Jeroen, can you please clarify what you mean by "about the period"? I believe this pattern was taken from call_perf(). (line 771 in portfolio_controller.py)

File: openbb_terminal/portfolio/portfolio_controller.py (hunk 4 above)
Comment (Sammac-dev) [reply openbb_terminal/portfolio/portfolio_controller.py:683]: For this we were using -r as an abbreviation for "--raw". Regarding the showing of the graph, based on what was mentioned earlier we were under the impression that "--raw" is only to show or disable the chart not the graph.

Please let us know if we have misunderstood the request and what behavior you would like the "-r"/"--raw" flag to exhibit.

image

File: openbb_terminal/portfolio/portfolio_controller.py (hunk 3 above)
Comment (JerBouma) [reply openbb_terminal/portfolio/portfolio_controller.py:664]: The `help` argument says "The file to be loaded" while this relates to the period. So the description is wrong.

File: openbb_terminal/portfolio/portfolio_controller.py (hunk 4 above)
Comment (JerBouma) [reply openbb_terminal/portfolio/portfolio_controller.py:683]: So, when you call `--raw`, I want to see the `table` and this is exactly what happens. However, the `help` argument says: `help="view attribution in a graph visualisation format"` and that's incorrect, you are actually activating the table by calling `--raw` not the graph so this is mostly a documentation thing.

**Comment by Sammac-dev** [comment]
Hi Jeroen, the requested changes have been made. Note that in order to push I had to pull first which updated the branch. "portfolio_controller.py" was the only file that I've updated but I'm unsure if there is an issue with the way I've committed (I'm fairly new to working with forks/branches)

**Review by JerBouma — APPROVED** [review]

**Pull request merged**